### PR TITLE
[Transform] Add basic onednn_graph dialect lowering

### PR DESCRIPTION
**Pull request by LongshengDu**
Added basic onednn_graph dialect lowering for matmul, relu, add.
Tracking: #28 
Depending on: #43

The lowering of elementwise binary ops with constant operand will use named ops, same as the other elementwise ops for consistency. And we will develop and explore its strength/limitation:
```
%cst = arith.constant dense<0.000000e+00> : tensor<128x256xf32>
%relued = linalg.max 
    ins(%biased, %cst : tensor<128x256xf32>, tensor<128x256xf32>) 
    outs(%output : tensor<128x256xf32>) -> tensor<128x256xf32>
```

We can explore more options in the future if named op cannot comply with our design, e.g.:
```
%c0f = arith.constant 0.0 : f32
%relued = linalg.elemwise_binary { fun = #linalg.binary_fn<max_signed> }
    ins(%biased, %c0f : tensor<512x512xf32>, f32)
    outs(%output : tensor<512x512xf32>) -> tensor<512x512xf32>
```

### CODE REVIEW

File: lib/gc/Transforms/OneDNNGraphToLinalg.cpp [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+          loc, DenseElementsAttr::get(ty, int64_t(I)));
+    } else if (llvm::isa<FloatType>(ty.getElementType())) {
+      return b.create<arith::ConstantOp>( //
+          loc, DenseElementsAttr::get(ty, float(I)));
+    } else {
+      op->emitError("Not a supported element type for constant.\n");
+      return nullptr;
+    }
+  }
+};
```
Comment (kurapov-peter) [inline lib/gc/Transforms/OneDNNGraphToLinalg.cpp:101]: These methods are stateless. What are they wrapped into a struct for? Is it just for logical grouping expecting more methods?

**Review by kurapov-peter — APPROVED** [review]
Just a couple minor comments, otherwise looks fine. I'd probably also add some failure test cases (e.g., for unexpected broadcasting) as a nice-to-have thing; not necessary in this PR though.

File: lib/gc/Transforms/OneDNNGraphToLinalg.cpp [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+          /*resultTensorTypes=*/resultTy,
+          /*inputs=*/ValueRange{op.getInputA(), op.getInputB()},
+          /*outputs=*/outTensor);
+    } else {
+      // T(B * A)
+      int64_t rank = resultTy.getRank();
+      SmallVector<int64_t> permutation(rank);
+      std::iota(std::begin(permutation), std::end(permutation), 0);
+      permutation[rank - 2] = rank - 1;
+      permutation[rank - 1] = rank - 2;
```
[the remark this answers is not in the document]
Comment (LongshengDu) [reply lib/gc/Transforms/OneDNNGraphToLinalg.cpp]: Yeah, for 2D matmul it should be static

File: lib/gc/Transforms/OneDNNGraphToLinalg.cpp [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+            /*location=*/loc,
+            /*inputs=*/op,
+            /*inits=*/initTensor,
+            /*dimensions=*/bcastDims)
+        .getResults()
+        .front();
+  }
+}
+
+typedef Value (*GetOperandFn)(Operation *, PatternRewriter &, TensorType);
```
[the remark this answers is not in the document]
Comment (LongshengDu) [reply lib/gc/Transforms/OneDNNGraphToLinalg.cpp]: This is a type define for a function that returns `mlir::Value`.

File: lib/gc/Transforms/OneDNNGraphToLinalg.cpp (hunk 1 above)
Comment (LongshengDu) [reply lib/gc/Transforms/OneDNNGraphToLinalg.cpp]: Just for logical grouping, will add more, e.g. get constant operand from attrs like `onednn_graph.pow`

**Review by kurapov-peter — APPROVED** [review]
Looks like a reasonable starting point now

**Pull request merged**